clic.adoc

```diff
@@ -1720,49 +1720,37 @@ address.
 [source]
 ----
 /* MRET pseudo-code */
-set_next_pc(exception_handler(cur_privilege, MRET, PC));
-
 function exception_handler(cur_priv, xret, pc) {
   match (xret) {
 ...
-      MRET =>  {
-      let xepc = prepare_xret_target(cur_priv);
-      let xinhv = prepare_xinhv_target(cur_priv);
-
+    MRET =>  {
+      /* standard MRET behavior */
       let prev_priv = cur_priv;
       mstatus.MIE   = mstatus.MPIE;
       mstatus.MPIE  = 1;
       cur_priv      = mstatus.MPP;
       ... /* additional standard MRET behavior */
 
-      if xinhv 
-      then {  
-        if (check_fetch_permissions(xepc) = Addr_OK)
-            next_pc = mem_read(xepc) & ~1; /* xepc contains an address of a table entry */
-          } else {
+      if (minhv) {
+        let mepc_inhv = mepc & inhv_pc_alignment_mask(); /* align xepc value to XLEN/8 byte boundary by ignoring low-order bits. */
+        if (check_fetch_permissions(mepc_ihnv) = Addr_OK) {
+            next_pc = mem_read(mepc_inhv) & ~1; /* mepc_inhv contains an address of a table entry */
+        } else {
             /* take table-fetch trap */
-          }
-        } else { /* Standard xRET behavior - xepc becomes next_pc */
-          next_pc = xepc & ~1;
+            mcause.minhv = 1;
+            /* other side effects of taking a trap */
+        }
+      } else { /* Standard MRET behavior - mepc becomes next_pc */
+        if (check_fetch_permissions(mepc) = Addr_OK) {
```

Comment:
What does check_fetch_permissions do? Checking for example PMP permissions is not part of the standard MRET behavior, right? That should be done as part of the (possibly failing) instruction fetch at the target address.